cdap-security/src/main/java/io/cdap/cdap/security/tools/AccessTokenGenerator.java

```diff
@@ -0,0 +1,220 @@
+      new ConfigModule(),
+      new IOModule(),
+      new AbstractModule() {
+        @Override
+        protected void configure() {
+          // We don't do anything with metrics, so just bind to the no-op implementation
+          bind(MetricsCollectionService.class).to(NoOpMetricsCollectionService.class);
+        }
+      },
+      CoreSecurityRuntimeModule.getDistributedModule(cConf));
+    tokenManager = injector.getInstance(TokenManager.class);
+    tokenCodec = injector.getInstance(Key.get(new TypeLiteral<Codec<AccessToken>>() { }));
+
+    CommonNettyHttpServiceFactory factory = injector.getInstance(CommonNettyHttpServiceFactory.class);
+    NettyHttpService.Builder builder = factory.builder("accesstoken.generator")
+      .setHost(host)
+      .setPort(port)
+      .setExecThreadPoolSize(execThreadPoolSize)
+      .setWorkerThreadPoolSize(workerThreadPoolSize)
+      .setHttpHandlers(new TokenGeneratorHttpHandler(cConf, tokenManager, tokenCodec, completion));
+    if (cConf.getBoolean(Constants.Security.SSL.INTERNAL_ENABLED)) {
+      SConfiguration sConf = injector.getInstance(SConfiguration.class);
+      new HttpsEnabler().configureKeyStore(cConf, sConf).enable(builder);
+    }
+    httpService = builder.build();
+  }
+
+  @VisibleForTesting
+  NettyHttpService getService() {
+    return httpService;
+  }
+
+  @VisibleForTesting
+  TokenManager getTokenManager() {
+    return tokenManager;
+  }
+
+  @VisibleForTesting
+  Codec<AccessToken> getTokenCodec() {
+    return tokenCodec;
+  }
+
+  @Path(Constants.Gateway.INTERNAL_API_VERSION_3 + "/accesstoken")
+  public static class TokenGeneratorHttpHandler extends AbstractHttpHandler {
+    private final CConfiguration cConf;
+    private final TokenManager tokenManager;
+    private final Codec<AccessToken> tokenCodec;
+    private final CompletableFuture<?> completion;
+
+    public TokenGeneratorHttpHandler(CConfiguration cConf, TokenManager tokenManager,
+                                     Codec<AccessToken> tokenCodec, CompletableFuture<?> completion) {
+      this.cConf = cConf;
+      this.tokenManager = tokenManager;
+      this.tokenCodec = tokenCodec;
+      this.completion = completion;
+    }
+
+    @Override
+    public void init(HandlerContext context) {
+      tokenManager.startUp();
+    }
+
+    @Override
+    public void destroy(HandlerContext context) {
+      tokenManager.stopAndWait();
+      completion.complete(null);
+    }
+
+    @GET
+    @Path("/")
+    public void getToken(FullHttpRequest request, HttpResponder responder,
+                         @QueryParam("username") String username,
+                         @QueryParam("identifierType") @DefaultValue("INTERNAL") String identifierType)
+      throws IOException {
+      UserIdentity.IdentifierType type;
+      try {
+        type = UserIdentity.IdentifierType.valueOf(identifierType);
+      } catch (IllegalArgumentException e) {
+        responder.sendString(HttpResponseStatus.BAD_REQUEST,
```

Comment:
Use the `HttpExceptionHandler` in the service and throw a BadRequestException from here instead.

Reply:
fixed